lucet-runtime/lucet-runtime-internals/src/instance.rs

```diff
@@ -562,19 +561,21 @@ impl Instance {
     /// The foreign code safety caveat of [`Instance::run()`](struct.Instance.html#method.run)
     /// applies.
     pub fn resume_with_val<A: Any + 'static>(&mut self, val: A) -> Result<RunResult, Error> {
-        Ok(self.resume_with_val_impl(val, false, None)?.unwrap())
+        Ok(self
+            .resume_with_val_impl(Box::new(val), None, None)?
+            .unwrap())
     }
 
-    pub(crate) fn resume_with_val_impl<A: Any + 'static>(
+    pub(crate) fn resume_with_val_impl(
         &mut self,
-        val: A,
-        async_context: bool,
+        val: Box<dyn Any + 'static>,
```

Comment:
Mostly, this allows resuming the instance without double-boxing in the `RunAsync` future. The type of a value generally has to be known statically for a box to be allocated, so without this, we'd need to heap allocate twice (`Box<Box<ResumeVal>>`).

There's also no additional type check: because the expected type of the resume value is not known statically, the compiler can't remove the check in the original version. Further, the implementation of `dyn Any::is` is not const, and the implementation is nearly the same:

https://github.com/rust-lang/rust/blob/8fe989dd768f5dfdb0fc90933f3f74fa4579fefd/library/core/src/any.rs#L184-L195

Even then, the cost of a branch should be far less than a heap allocation. Especially because we're checking a developer invariant, branch misses should never happen in correct code. It might be worth adding `#[unlikely]` / the stable equivalent, but also, a branch miss is very little compared to the cost of re-entering an instance, and probably isn't worth the added code complexity.

Reply:
Also, the address of the typeid of `dyn A` is at a statically known location.

Reply:
OK, yes, all of this makes sense -- thanks for the detailed explanation!